### ISSUE TITLE
CookBook.ipynb

### ISSUE THREAD

**Issue by itrus** [andersen-lab/paper_2018_primalseq-ivar#1]
Hello,

I have a few suggestions for the CookBook (CookBook.ipynb).

1. "In [11]" will not work in the absence of an index. The followoing line could be added
`bwa index PRV.fa `
2. "In [13]" "ivar trim" trims (deletes) 100% reads in your output. Most probably because BED file was missing. This is not crucial to reproduce your pathway. However, continuing after deleting all reads (going to the next steps) is impossible.
```
Trimmed primers from 644979 reads.
69219 reads were shortened below the minimum length of 30 bp and were not written to file.
576690 reads that started outside of primer regions were not written to file.
```
3. "In [32]" output contains the following line. Most probably the proper BAM file was already there. And an additional one was not created.
`[bam_merge] File 'Z52.merged.bam' exists. Please apply '-f' to overwrite. Abort.`
4. The following line is duplicated both in "In [31]" and "In [32]". One command should be enough.
`bwa mem -k 5 -T 16 db/Z52.consensus db/zika_primers.fa | samtools view -bS -F 4 -o db/zika_primers_consensus.bam`
5. "In [42]" always returns me the following error. Both for your (ZI52) and my files. I don't know awk enough,   but most probably it is something going with this file: "get_common_variants.sh". Please advise.
```
awk: line 6: illegal reference to array FILES
awk: line 30: illegal reference to array FILES
awk: line 33: illegal reference to array line_arr

```

Everything else is great! Thank you for the nice package.

**Comment by gkarthik** [comment]
Hey Ivan, thank you for the suggestions! I made the following changes and pushed a commit.

1. I added a bwa index step before alignment. 
2. The bed files is in the [db folder](https://github.com/andersen-lab/paper_2018_primalseq-ivar/blob/master/cookbook/db/). Not supplying it will remove all reads without primers sequence and like you said will remove all of the reads since there are no primers supplied. If you want to keep the reads without primers, I've added an `-e` flag to the `trim` command.
`ivar trim -e -i <input.bam> -b <primers.bed> -p`
3. I added a comment in that cell to include the `-f` flag when re-running the command.
4. I removed the duplicate lines. However, in this step it is essential to sort the output BAM before calling the mpileup. SO, I edited the line to,
`bwa mem -k 5 -T 16 db/Z52.consensus db/zika_primers.fa | samtools view -bS -F 4 | samtools sort -o db/zika_primers_consensus.bam`
5. Can you re run the notebook now? I wasn't able to reproduce this error. If the problem persists could you include information on your **OS**?

**Comment by itrus** [comment]
Hey Karthik,

Thank you very much for taking care of this issue! Finally, I think that this error message depends on the OS.
I  was testing it in Debian-based OS (Ubuntu (1 computer) and Kubuntu (2 different computers)). The error was always there. While with Arch-based OS (Manjaro) "ivar filtervariants" was working as expected. Thus, you are right that this is an OS-specific issue.

Ivan

**Comment by gkarthik** [comment]
Weird. I tested this on bash in Ubuntu and it should work fine.
Did you test it in bash on Ubuntu and Kubuntu? It maybe that some shells like zsh might have a different syntax.

**Comment by itrus** [comment]
Thank you. You gave me a clue, Now I know what is going on. I have a recent BASH (5.0.3). But this is not important. It is important what is going on with awk. Manjaro is calling **gawk** each time when awk is invoked. While Ubuntu/Kubuntu by default has only implementation of awk with the **mawk** package.

I solved this issue in Kubuntu by running the next commands:

```
sudo apt-get install gawk
sudo update-alternatives --config awk
```

The first one installs gawk. The second one allows selecting gawk as the main interpreter of awk. However, the replacement happened automatically. Thus, I assume that the second command was not needed.

To sum it up:

1. "ivar filtervariants" is working well. I've checked its output.
2. I suggest to include this tip ("awk should be realized with gawk and nothing else") in the manual for iVAR.

Thank you for your kind assistance,
Ivan
